### PR TITLE
Handling GitHub API Rate Limit Exceeded Exception

### PR DESCRIPTION
**Pull request by idavidov**
PR Type:
**Enhancement**

___
PR Description:
**This PR introduces handling for the RateLimitExceededException from the GitHub API. It adds a retry mechanism for the get_diff_files method in the GithubProvider class and logs an error message when the rate limit is exceeded. Additionally, a new custom exception, RateLimitExceeded, is introduced in the servers/utils.py file. The PR also includes a new configuration setting for the number of retries when the rate limit is exceeded.**

___
PR Main Files Walkthrough:
-`pr_agent/algo/pr_processing.py`: Added exception handling for RateLimitExceededException when calling get_diff_files method from the git provider.
-`pr_agent/git_providers/github_provider.py`: Added a retry mechanism for the get_diff_files method and raised a custom exception, RateLimitExceeded, when the GitHub API rate limit is exceeded.
-`pr_agent/servers/utils.py`: Introduced a new custom exception, RateLimitExceeded.
-`pr_agent/settings/configuration.toml`: Added a new configuration setting for the number of retries when the rate limit is exceeded.

### CODE REVIEW

**Comment by okotek** [comment]
Thanks for your contribution @idavidov !
Do you mind adding @retry decorator that handles the rate limit exception, similar to the handling in ai_handler.py for chat_completion()?

**Comment by idavidov** [comment]
sure, i'll do

File: pr_agent/algo/pr_processing.py [hunk 1]
```diff
@@ -40,7 +45,11 @@ def get_pr_diff(git_provider: GitProvider, token_handler: TokenHandler, model: s
         global PATCH_EXTRA_LINES
         PATCH_EXTRA_LINES = 0
 
-    diff_files = list(git_provider.get_diff_files())
+    try:
+        diff_files = list(git_provider.get_diff_files())
+    except GithubException.RateLimitExceededException as e:
+        logging.error(f"Rate limit exceeded for GitHub API. original message {e}")
+        raise
```
[the remark this answers is not in the document]
Comment (idavidov) [reply pr_agent/algo/pr_processing.py:52]: raise added in order to support @retry decorator

**Comment by okotek** [comment]
I'd add it to configuration.toml
image

File: pr_agent/algo/pr_processing.py [hunk 2]
```diff
@@ -19,7 +21,10 @@
 OUTPUT_BUFFER_TOKENS_HARD_THRESHOLD = 600
 PATCH_EXTRA_LINES = 3
 
+GITHUB_RETRIES=1
 
+@retry(exceptions=(APIError, Timeout, TryAgain, AttributeError, RateLimitError, GithubException.RateLimitExceededException),
```
Comment (okotek) [inline pr_agent/algo/pr_processing.py:26]:
```suggestion
@retry(exceptions=(GithubException.RateLimitExceededException),
```

**Review by okotek — CHANGES_REQUESTED** [review]

**Comment by okotek** [comment]
@idavidov would you mind moving the retry mechanism to github_provider's get_diff_files() - we shouldn't reference a specific provider in pr_processing.py

**Comment by idavidov** [comment]
sure. make sense let me do it

**Comment by idavidov** [comment]
ok, no problem, let me see how do it i need gain some more familiarity with pr_agent structure, but i would love to do it

**Comment by okotek** [comment]
Let me know if you want me to take over

**Comment by idavidov** [comment]
i love this project i would like to learn more about, and make long term contribution
if it works for you i would love to progress with this one and take more in future

**Comment by idavidov** [comment]
Ori i think i got your idea and implemented it hopefully in right way
Please review and let me know if it is good
thank you for your patience
Ilya

**Review by okotek — DISMISSED** [review]

**Comment by okotek** [comment]
The new change makes the code not run due to cyclic dependency. Let me fix it

**Review by okotek — APPROVED** [review]

**Pull request merged**